lib/jellygrinder/client.ex

```diff
@@ -1,7 +1,115 @@
 defmodule Jellygrinder.Client do
   @moduledoc false
 
+  alias Jellygrinder.Client
+  alias Jellygrinder.Client.Helpers.ConnectionManager
+  alias Jellygrinder.Coordinator
+
   @type options() :: %{uri: URI.t(), name: String.t()}
 
   @callback start_link(options()) :: GenServer.on_start()
+
+  @doc false
+  defmacro __using__(_opts) do
+    quote do
+      def request(path, label, state, retries \\ 0) do
+        Client.request(path, label, state, retries)
+      end
+
+      def get_track_manifest_name(master_manifest) do
+        Client.get_track_manifest_name(master_manifest)
+      end
+
+      def trim_manifest(manifest, pattern) do
+        Client.trim_manifest(manifest, pattern)
+      end
+    end
+  end
+
+  @spec request(String.t(), String.t(), map(), non_neg_integer()) :: {:ok | :error, binary()}
+  def request(path, label, state, 0), do: request_and_log(path, label, state)
+
+  def request(path, label, state, retries) do
+    case request_and_log(path, label, state) do
+      {:ok, _content} = response -> response
+      {:error, _reason} -> request(path, label, state, retries - 1)
+    end
+  end
+
+  @spec get_track_manifest_name(String.t()) :: String.t()
+  def get_track_manifest_name(master_manifest) do
+    master_manifest
+    |> String.split("\n", trim: true)
+    |> List.last()
+  end
+
+  @doc """
+  Trim the manifest, returning everything after `pattern`
+  If `pattern == nil` or `pattern` isn't present in manifest, return the entire manifest
+  """
+  @spec trim_manifest(String.t(), String.t() | nil) :: String.t()
+  def trim_manifest(manifest, nil), do: manifest
+
+  def trim_manifest(manifest, pattern) do
+    manifest
+    |> String.split(pattern, parts: 2)
+    |> Enum.at(1, manifest)
+  end
```

Comment:
Why are these methods public? Shouldn't we inline them in `__using__` section? Will we call it without using this behavior earlier?

Reply:
I have read that this is a pattern. The `__using__` scope should be as simple as possible also if you put all the public functions inside the `__using__` section you have to do the same with private functions.